types/applications/formApplication.d.ts

```diff
@@ -60,7 +64,7 @@ declare abstract class FormApplication<O = {}> extends Application {
    * @param options - (unused) (default: `{}`)
    * @override
    */
-  getData(options?: Application.RenderOptions): FormApplication.Data<O> | Promise<FormApplication.Data<O>>;
+  getData(options?: Application.RenderOptions): D | Promise<D>;
```

Comment:
We could make the `Pormise` part of the type `D` itself. This has the benefit that deriving classes do not need to have both the promise and non promise variant as return type. But you need to "deprmosify" `D` for the parameter of `_renderInner` then:
`typesUtils.d.ts`:
```ts
type Awaited<T> = T extends PromiseLike<infer U> ? U : T;
```
`formApplication.d.ts`:
```ts
declare abstract class FormApplication<
  D extends object | Promise<object> = FormApplication.Data<{}>,
  O extends object = D extends FormApplication.Data<infer T> ? T : {}
> extends Application {
  /* ... */
  getData(options?: Application.RenderOptions): D;
  /* ... */
  protected _renderInner(data: Awaited<D>, options?: Application.RenderOptions): Promise<JQuery>;
  /* ... */
}
```

Reply:
I do think it would be nice to have the exact return type of `getData()` on the base implementation, but it could be messy. As it stands now, if a subclass specifies just `D` or just `Promise<D>`, it won't complain.

Reply:
didn't work for me, but I can check again